docs/source/migration/4.0.mdx

```diff
@@ -40,9 +48,18 @@ try {
 // With
 val response = client.query(MyQuery()).execute()
 if (response.exception != null) {
-  // Handle network error
-} else if (response.hasErrors()) {
-  // Handle GraphQL errors
+  when (response.exception) {
+    is ApolloGraphQLException -> {
+      // Handle GraphQL errors
+    }
+    else -> {
+      // Handle network error
+    }
+  }
+} else {
+  // No errors
+  val data = response.data
```

Comment:
What about doing the opposite and check `data` first so that it can be smart cast? :

```kotlin
val data = response.data
if (data ! = null) {
  // do something with data here
} else {
  // do something with exception here
}
```

Reply:
👍 flipped it